catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/AccessControlPreQueryPlugin.java

```diff
@@ -0,0 +1,237 @@
+      return;
+    }
+    if (!(tag instanceof String)) {
+      LOGGER.error(
+          "The service property 'access-controlled-tag' is not valid, verify blueprint is correct");
+      return;
+    }
+    accessControlledTags.remove(tag);
+  }
+
+  @Override
+  public QueryRequest process(QueryRequest input)
+      throws PluginExecutionException, StopProcessingException {
+    final String subjectIdentifier = getSubjectIdentifier();
+    final Set<String> groups = new HashSet<>(getSubjectGroups());
+    final String groupThatCanSeeEverything = configuration.getSystemUserAttributeValue();
+
+    if (groups.contains(groupThatCanSeeEverything)) {
+      LOGGER.debug(
+          "Will not modify filter; subject ({}) had at least one group [{}] that was exempt [{}]",
+          subjectIdentifier,
+          groups,
+          groupThatCanSeeEverything);
+      return input;
+    }
+
+    final TagAggregationVisitor tagVisitor = new TagAggregationVisitor();
+    final SimplifyingFilterVisitor simpleVisitor = new SimplifyingFilterVisitor();
+
+    final Query query = input.getQuery();
+    final Filter simplified = (Filter) query.accept(simpleVisitor, null);
+
+    simplified.accept(tagVisitor, null);
+
+    final Set<String> discoveredTags = tagVisitor.getTags();
+    if (CollectionUtils.isEmpty(discoveredTags)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] did not completely bind results to tags",
+          subjectIdentifier,
+          query);
+      return input;
+    }
+
+    Set<String> tagsNotInCollaborationSet =
+        Sets.filter(discoveredTags, tag -> !accessControlledTags.contains(tag));
+    if (CollectionUtils.isNotEmpty(tagsNotInCollaborationSet)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] referenced tags not in collab set [{}]",
+          subjectIdentifier,
+          query,
+          accessControlledTags);
+      return input;
+    }
+
+    Filter policyBranch = createBakedInSecurityPolicySubset(subjectIdentifier, groups);
+    Filter combined = filterBuilder.allOf(query, policyBranch);
+    return new QueryRequestImpl(
+        new QueryImpl(combined), input.isEnterprise(), input.getSourceIds(), input.getProperties());
+  }
+
+  @VisibleForTesting
+  String getSubjectIdentifier() {
+    final Subject subject = (Subject) SecurityUtils.getSubject();
+    final String uniqueIdentifier = identity.getUniqueIdentifier(subject);
+    LOGGER.debug("Obtained user's unique identifier: {}", uniqueIdentifier);
+    return uniqueIdentifier;
+  }
+
+  @VisibleForTesting
+  List<String> getSubjectGroups() {
+    final Subject subject = (Subject) SecurityUtils.getSubject();
+    final List<String> groups =
+        SubjectUtils.getAttribute(subject, configuration.getSystemUserAttribute());
+    LOGGER.debug("Obtained user's groups: {}", groups);
+    return groups;
+  }
+
+  private Filter createBakedInSecurityPolicySubset(String identifier, Set<String> groups) {
```

Comment:
What does `BakedIn` vs non-`BakedIn` mean? I assume this could just be `createSecurityPolicySubet`

Reply:
Correct. I'll update.